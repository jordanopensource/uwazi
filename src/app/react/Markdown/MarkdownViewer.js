import PropTypes from 'prop-types';
import React, { Component } from 'react';
import sanitizeHtml from 'sanitize-html';
import { risonDecodeOrIgnore } from 'app/utils';
import { Translate } from 'app/I18N';
import { MarkdownLink, SearchBox, MarkdownMedia, ItemList } from './components';
import CustomHookComponents from './CustomHooks';

import markdownToReact from './markdownToReact';
import { ValidatedElement } from './ValidatedElement';
import { visualizationHtmlTags } from './utils';

class MarkdownViewer extends Component {
  static errorHtml(index, message) {
    return (
      <p key={index} className="error">
        <br />
        <strong>
          <i>
            <Translate translationKey="Custom component error">
              Custom component markup error: unsupported values! Please check your configuration
            </Translate>
          </i>
        </strong>
        <br />
        {message}
        <br />
      </p>
    );
  }

  static customHook(config, index) {
    const props = risonDecodeOrIgnore(config);
    if (!CustomHookComponents[props.component]) {
      throw new Error('Invalid  component');
    }
    const Element = CustomHookComponents[props.component];
    return <Element {...props} key={index} />;
  }

  inlineComponent(type, config, index) {
    const { compact } = this.props;
    let result;
    if (type === 'list') {
      result = this.list(config, index);
    }

    if (type === 'link') {
      result = <MarkdownLink {...risonDecodeOrIgnore(config)} key={index} />;
    }

    if (type === 'searchbox') {
      result = <SearchBox {...risonDecodeOrIgnore(config)} key={index} />;
    }

    if (['vimeo', 'youtube', 'media'].includes(type)) {
      result = (
        <div key={index}>
          <MarkdownMedia key={index} config={config} compact={compact} />
        </div>
      );
    }

    if (type === 'customhook') {
      result = MarkdownViewer.customHook(config, index);
    }
    return result;
  }

  customComponent(type, config, index, children) {
    try {
      if (typeof type === 'function' || typeof type === 'object') {
        const Element = type;
        return (
          <Element {...config} key={index}>
            {children}
          </Element>
        );
      }

      if (type) {
        return this.inlineComponent(type, config, index);
      }
    } catch (error) {
      return MarkdownViewer.errorHtml(index, error.message);
    }

    return false;
  }

  list(_config, index) {
    const listData = this.props.lists[this.renderedLists] || {};
    const output = (
      <ItemList
        key={index}
        link={`/library/${listData.params}`}
        items={listData.items}
        options={listData.options}
      />
    );
    this.renderedLists += 1;
    return output;
  }

  render() {
    this.renderedLists = 0;

    if (typeof this.props.markdown !== 'string') {
      return false;
    }

    const sanitizedMarkdown = !this.props.sanitized
      ? this.props.markdown
      : sanitizeHtml(this.props.markdown, {
          allowedTags: visualizationHtmlTags,
          allowedAttributes: false,
        });

    const ReactFromMarkdown = markdownToReact(
      sanitizedMarkdown,
      this.customComponent.bind(this),
      this.props.html
    );

    if (!ReactFromMarkdown) {
      return false;
    }

    return ValidatedElement(
      'div',
      { className: 'markdown-viewer' },
      React.Children.toArray(ReactFromMarkdown),
      this.props.sanitized
    );
  }
}

MarkdownViewer.defaultProps = {
  lists: [],
  markdown: '',
  html: false,
  compact: false,
  sanitized: true,
};

MarkdownViewer.propTypes = {
  markdown: PropTypes.string,
  lists: PropTypes.arrayOf(PropTypes.object),
  html: PropTypes.bool,
  compact: PropTypes.bool,
  sanitized: PropTypes.bool,
};

export default MarkdownViewer;
