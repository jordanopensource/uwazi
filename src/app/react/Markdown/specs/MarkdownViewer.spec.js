/* eslint-disable no-template-curly-in-string */
/* eslint-disable max-statements */
import React, { Component } from 'react';

import { shallow } from 'enzyme';

import CustomHookComponents from '../CustomHooks';
import MarkdownViewer from '../MarkdownViewer';

describe('MarkdownViewer', () => {
  let component;
  let props;

  beforeEach(() => {
    props = {
      markdown: '## MarkdownContent',
      sanitized: false,
    };
  });

  const render = () => {
    component = shallow(<MarkdownViewer {...props} />);
  };

  const matchSnapshot = () => {
    render();
    expect(component).toMatchSnapshot();
  };

  describe('render', () => {
    describe('when markdown is invalid', () => {
      it('should not render anything when its empty', () => {
        props.markdown = '';
        render();

        expect(component).toMatchSnapshot();
      });

      it('should not render when its not a string', () => {
        props.markdown = [];
        render();

        expect(component.isEmptyRender()).toBe(true);
      });
    });

    it('should render markdown', () => {
      props.markdown = `## title\nSome text with a [URL](http://google.com) inside.
        \n\nWhich should be in its own line, separated with TWO line breaks (to create a new <p> Element)
        \n\n\`\`\`javascript\nCode\n\`\`\`
      `;

      matchSnapshot();
    });

    it('should remove whitespaces between table tags (to prevent react warning)', () => {
      props.markdown = `
| cadh | cbdp | cidfp | cipst |
| - | - | - | - |
| 1.1,25.1 |  |  |  |
| 1.1,21.1,21.2,25,1 |  |  |  |
      `;

      matchSnapshot();
    });

    it('should support containers with custom classNames', () => {
      props.markdown = '::: test classname \ntext inside a div';

      matchSnapshot();
    });

    it('should render media components', () => {
      props.markdown = `{vimeo}(1234, options)
      \n\n{youtube}(1234)
      \n\n{media}(config options)
      `;

      matchSnapshot();
    });

    it('should render list components', () => {
      props.lists = [
        { params: 'param1', items: ['items1'], options: { option1: 'optionValue' } },
        { params: 'param2', items: ['items2'] },
        { params: 'param3', items: ['items3'] },
      ];

      props.markdown = `{list}(first list)
        \n\n{list}(http://secondList)(limit:3)
        \n\n{list}(asd)
      `;

      render();
      expect(component).toMatchSnapshot();
    });

    it('should render customHook components and show an error for invalid ones', () => {
      CustomHookComponents.validcomponent = class customComponent extends Component {
        render() {
          return <p>Custom</p>;
        }
      };

      props.markdown = `
        \n\n{customhook}(component:validcomponent,prop:'a prop',array:!((child:!(a),bool:!t),(child:!(b))))
        \n\nShould allow markdown between hooks
        \n\n{customhook}(component:validcomponent)
        \n\n{customhook}(component:invalidcomponent)
      `;

      render();
      expect(component).toMatchSnapshot();
    });

    it('should be able to render properly custom components not separated by \\n\\n', () => {
      props.markdown = '{youtube}(youtubeUrl)\n{media}(mediaUrl)';

      render();
      expect(component).toMatchSnapshot();
    });

    it('should not render html by default', () => {
      props.markdown = '<div><h1>should be all a escaped string</h1></div>';

      render();
      expect(component).toMatchSnapshot();
    });

    it('should render Link', () => {
      props.markdown =
        "{link}(url:this_is_the_url,label:'this is a label')" +
        '\n<MarkdownLink url="the_url">label</MarkDownLink>';
      props.html = true;

      render();
      expect(component).toMatchSnapshot();
    });

    describe('when passing html true prop', () => {
      it('should render html', () => {
        props.html = true;
        props.markdown = '<div><h1>test</h1></div>';

        render();
        expect(component).toMatchSnapshot();
      });

      it('should render customComponents nested inside html', () => {
        props.html = true;
        props.lists = [
          { params: 'param1', items: ['items1'], options: { option1: 'optionValue' } },
        ];
        props.markdown = '<div>{list}(url)(options) <div>{youtube}(url)</div></div>';

        render();
        expect(component).toMatchSnapshot();
      });
    });

    describe('when not valid html', () => {
      it('should not fail on malformed tags', () => {
        props.markdown = '<div><h</div>';
        props.html = true;

        render();
        expect(component).toMatchSnapshot();
      });

      it('should not fail on unsupported tags', () => {
        props.html = true;
        props.markdown =
          "<b>Little red ridding hood</b>\n<p>I don't know <..long pause..> a minute later  <..pause..> a grandma story.\nWhen I heard it</p>";

        render();
        expect(component).toMatchSnapshot();
      });
    });

    it('should render a searchbox', () => {
      props.markdown = "{searchbox}(placeholder:'this is a placeholder')";

      render();
      expect(component).toMatchSnapshot();
    });

    it('should render single tags as self closing', () => {
      props.html = true;
      props.markdown = 'test\n\n***\ntest';

      render();
      expect(component).toMatchSnapshot();
    });

    it('should render properly a selfclosing XML tags', () => {
      props.html = true;
      props.markdown = 'test\n\n<SearchBox/>\n<div>test</div>';

      render();
      expect(component).toMatchSnapshot();
    });

    it('should remove Dataset and Query tags', () => {
      props.html = true;
      props.markdown =
        '' +
        'test\n\n' +
        '<Dataset />\n' +
        '<Query />\n' +
        '<SearchBox/>\n' +
        '<Dataset name="test" />\n' +
        '<div>test</div>';

      render();
      expect(component).toMatchSnapshot();
    });
  });

  describe('limited markdown', () => {
    it.each`
      type                | markdown
      ${'forbidden tag'}  | ${'<div>This <input type="text" /> was cleaned</div>'}
      ${'forbidden tag'}  | ${'<div>This <script language="javascript">console.log(\'dangerous\')</script> was cleaned</div>'}
      ${'malicious code'} | ${'<img src="javascript:alert(\'forbidden\');" />'}
    `('should replace banned tags $type', ({ markdown }) => {
      props = {
        markdown,
        html: true,
        sanitized: true,
      };
      render();

      expect(component).toMatchSnapshot();
    });

    it.each`
      type               | markdown
      ${'custom tags'}   | ${'<MarkdownLink url="the_url">label</MarkDownLink> and <SearchBox/>\n<div>test</div>'}
      ${'custom hook'}   | ${'<CejilChart002 />'}
      ${'placeholder'}   | ${'<placeholder>$content</placeholder>'}
      ${'standard tags'} | ${'<Table><tr><th>Title</th></tr><tr><td>value</td></tr> </Table>'}
      ${'interpolation'} | ${'${template.color}'}
      ${'media'}         | ${'{vimeo}(1234, options)'}
    `('should keep allowed tags $type', ({ markdown }) => {
      props = {
        markdown,
        html: true,
        sanitized: true,
      };
      render();

      expect(component).toMatchSnapshot();
    });
  });
});
