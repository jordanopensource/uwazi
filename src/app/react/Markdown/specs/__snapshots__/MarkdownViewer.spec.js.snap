// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownViewer limited markdown should keep allowed tags custom hook 1`] = `
<div
  className="markdown-viewer"
>
  <p>
    <cejilchart002 />
  </p>
  

</div>
`;

exports[`MarkdownViewer limited markdown should keep allowed tags custom tags 1`] = `
<div
  className="markdown-viewer"
>
  <p>
    <MarkdownLink
      classname=""
      url="the_url"
    >
      label
    </MarkdownLink>
     and
  </p>
  

  <div>
    test
  </div>
</div>
`;

exports[`MarkdownViewer limited markdown should keep allowed tags interpolation 1`] = `
<div
  className="markdown-viewer"
>
  <p>
    \${template.color}
  </p>
  

</div>
`;

exports[`MarkdownViewer limited markdown should keep allowed tags media 1`] = `
<div
  className="markdown-viewer"
>
  <div>
    <MarkdownMedia
      compact={false}
      config="(1234, options)"
    />
  </div>
  

</div>
`;

exports[`MarkdownViewer limited markdown should keep allowed tags placeholder 1`] = `
<div
  className="markdown-viewer"
>
  <p>
    <placeholder>
      $content
    </placeholder>
  </p>
  

</div>
`;

exports[`MarkdownViewer limited markdown should keep allowed tags standard tags 1`] = `
<div
  className="markdown-viewer"
>
  <table>
    <tr>
      <th>
        Title
      </th>
    </tr>
    <tr>
      <td>
        value
      </td>
    </tr>
  </table>
</div>
`;

exports[`MarkdownViewer limited markdown should replace banned tags forbidden tag 1`] = `
<div
  className="markdown-viewer"
>
  <div>
    This  was cleaned
  </div>
</div>
`;

exports[`MarkdownViewer limited markdown should replace banned tags forbidden tag 2`] = `
<div
  className="markdown-viewer"
>
  <div>
    This  was cleaned
  </div>
</div>
`;

exports[`MarkdownViewer limited markdown should replace banned tags malicious code 1`] = `
<div
  className="markdown-viewer"
>
  <img />
</div>
`;

exports[`MarkdownViewer render should be able to render properly custom components not separated by \\n\\n 1`] = `
<div
  className="markdown-viewer"
>
  <div>
    <MarkdownMedia
      compact={false}
      config="(youtubeUrl)"
    />
  </div>
  

  <div>
    <MarkdownMedia
      compact={false}
      config="(mediaUrl)"
    />
  </div>
  

</div>
`;

exports[`MarkdownViewer render should not render html by default 1`] = `
<div
  className="markdown-viewer"
>
  <p>
    &lt;div&gt;&lt;h1&gt;should be all a escaped string&lt;/h1&gt;&lt;/div&gt;
  </p>
  

</div>
`;

exports[`MarkdownViewer render should remove Dataset and Query tags 1`] = `
<div
  className="markdown-viewer"
>
  <p>
    test
  </p>
  

  

  

  <SearchBox
    classname=""
    placeholder=""
  />
  

  

  <div>
    test
  </div>
</div>
`;

exports[`MarkdownViewer render should remove whitespaces between table tags (to prevent react warning) 1`] = `
<div
  className="markdown-viewer"
>
  <table>
    <thead>
      <tr>
        <th>
          cadh
        </th>
        <th>
          cbdp
        </th>
        <th>
          cidfp
        </th>
        <th>
          cipst
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          1.1,25.1
        </td>
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td>
          1.1,21.1,21.2,25,1
        </td>
        <td />
        <td />
        <td />
      </tr>
    </tbody>
  </table>
  

</div>
`;

exports[`MarkdownViewer render should render Link 1`] = `
<div
  className="markdown-viewer"
>
  <MarkdownLink
    classname=""
    label="this is a label"
    url="this_is_the_url"
  />
  

  <p>
    <MarkdownLink
      classname=""
      url="the_url"
    >
      label
    </MarkdownLink>
  </p>
  

</div>
`;

exports[`MarkdownViewer render should render a searchbox 1`] = `
<div
  className="markdown-viewer"
>
  <SearchBox
    classname=""
    placeholder="this is a placeholder"
  />
  

</div>
`;

exports[`MarkdownViewer render should render customHook components and show an error for invalid ones 1`] = `
<div
  className="markdown-viewer"
>
  <customComponent
    array={
      Array [
        Object {
          "bool": true,
          "child": Array [
            "a",
          ],
        },
        Object {
          "child": Array [
            "b",
          ],
        },
      ]
    }
    component="validcomponent"
    prop="a prop"
  />
  

  <p>
    Should allow markdown between hooks
  </p>
  

  <customComponent
    component="validcomponent"
  />
  

  <p
    className="error"
  >
    <br />
    <strong>
      <i>
        <Connect(Translate)
          translationKey="Custom component error"
        >
          Custom component markup error: unsupported values! Please check your configuration
        </Connect(Translate)>
      </i>
    </strong>
    <br />
    Invalid  component
    <br />
  </p>
  

</div>
`;

exports[`MarkdownViewer render should render list components 1`] = `
<div
  className="markdown-viewer"
>
  <ItemList
    items={
      Array [
        "items1",
      ]
    }
    link="/library/param1"
    options={
      Object {
        "option1": "optionValue",
      }
    }
  />
  

  <ItemList
    items={
      Array [
        "items2",
      ]
    }
    link="/library/param2"
    options={Object {}}
  />
  

  <ItemList
    items={
      Array [
        "items3",
      ]
    }
    link="/library/param3"
    options={Object {}}
  />
  

</div>
`;

exports[`MarkdownViewer render should render markdown 1`] = `
<div
  className="markdown-viewer"
>
  <h2>
    title
  </h2>
  

  <p>
    Some text with a 
    <a
      href="http://google.com"
    >
      URL
    </a>
     inside.
  </p>
  

  <p>
    Which should be in its own line, separated with TWO line breaks (to create a new &lt;p&gt; Element)
  </p>
  

  <pre>
    <code
      className="language-javascript"
    >
      Code

    </code>
  </pre>
  

</div>
`;

exports[`MarkdownViewer render should render media components 1`] = `
<div
  className="markdown-viewer"
>
  <div>
    <MarkdownMedia
      compact={false}
      config="(1234, options)"
    />
  </div>
  

  <div>
    <MarkdownMedia
      compact={false}
      config="(1234)"
    />
  </div>
  

  <div>
    <MarkdownMedia
      compact={false}
      config="(config options)"
    />
  </div>
  

</div>
`;

exports[`MarkdownViewer render should render properly a selfclosing XML tags 1`] = `
<div
  className="markdown-viewer"
>
  <p>
    test
  </p>
  

  <SearchBox
    classname=""
    placeholder=""
  />
  

  <div>
    test
  </div>
</div>
`;

exports[`MarkdownViewer render should render single tags as self closing 1`] = `
<div
  className="markdown-viewer"
>
  <p>
    test
  </p>
  

  <hr />
  

  <p>
    test
  </p>
  

</div>
`;

exports[`MarkdownViewer render should support containers with custom classNames 1`] = `
<div
  className="markdown-viewer"
>
  <div
    className="test classname"
  >
    <p>
      text inside a div
    </p>
    

  </div>
</div>
`;

exports[`MarkdownViewer render when markdown is invalid should not render anything when its empty 1`] = `""`;

exports[`MarkdownViewer render when not valid html should not fail on malformed tags 1`] = `
<div
  className="markdown-viewer"
>
  <div />
</div>
`;

exports[`MarkdownViewer render when not valid html should not fail on unsupported tags 1`] = `
<div
  className="markdown-viewer"
>
  <p>
    <b>
      Little red ridding hood
    </b>
  </p>
  

  <p>
    I don't know 
    <div
      className="error"
    >
      Invalid tag: ..long
    </div>
  </p>
</div>
`;

exports[`MarkdownViewer render when passing html true prop should render customComponents nested inside html 1`] = `
<div
  className="markdown-viewer"
>
  <div>
    <ItemList
      items={
        Array [
          "items1",
        ]
      }
      link="/library/param1"
      options={
        Object {
          "option1": "optionValue",
        }
      }
    />
    <div>
      <div>
        <MarkdownMedia
          compact={false}
          config="(url)"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`MarkdownViewer render when passing html true prop should render html 1`] = `
<div
  className="markdown-viewer"
>
  <div>
    <h1>
      test
    </h1>
  </div>
</div>
`;
